### PR TITLE
Implement Egg

### PR DESCRIPTION
**Pull request by Aaron1011**
This PR implements `Egg`, and creates a foundation for other projectiles that extend `EntityThrowable` in NMS.

### CODE REVIEW

File: src/main/java/org/spongepowered/mod/mixin/entity/projectile/MixinEntityEgg.java [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+@NonnullByDefault
+@Mixin(EntityEgg.class)
+public abstract class MixinEntityEgg extends EntityThrowable implements Egg {
+    public MixinEntityEgg(World worldIn) {
+        super(worldIn);
+    }
+
+    public double damageAmount;
+
+    @ModifyArg(method = "onImpact(Lnet/minecraft/util/MovingObjectPosition;)V", at =
```
Comment (Aaron1011) [inline src/main/java/org/spongepowered/mod/mixin/entity/projectile/MixinEntityEgg.java]: `onImpact` is an abstract method in `EntityThrowable`. Although it's always implemeted with a call to `attackEntityFrom`, it cannot be injected in `EntityThrowable`, as the bytecode is only present in a subclass.

File: src/main/java/org/spongepowered/mod/mixin/entity/projectile/MixinEntityEgg.java [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+        @At(value = "INVOKE", target="Lnet/minecraft/entity/Entity;attackEntityFrom(Lnet/minecraft/util/DamageSource;F)Z"))
+    private float onAttackEntityFrom(float damage) {
+        System.out.println("Hai");
```
Comment (Zidane) [inline src/main/java/org/spongepowered/mod/mixin/entity/projectile/MixinEntityEgg.java]: Forgot a printout :p?

Comment (Aaron1011) [inline src/main/java/org/spongepowered/mod/mixin/entity/projectile/MixinEntityEgg.java]: Oops :D

File: src/main/java/org/spongepowered/mod/mixin/entity/projectile/MixinEntityThrowable.java [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+
+    @Shadow
+    private String throwerName;
+
+    @Nullable
+    public ProjectileSource projectileSource;
+
+    @Override
+    public ProjectileSource getShooter() {
+        if (this.projectileSource != null && this.projectileSource instanceof ProjectileSource) {
```
Comment (simon816) [inline src/main/java/org/spongepowered/mod/mixin/entity/projectile/MixinEntityThrowable.java]: No need for `instanceof` checks here and below.

Comment (Aaron1011) [inline src/main/java/org/spongepowered/mod/mixin/entity/projectile/MixinEntityThrowable.java]: The `instanceof` check is actually needed for `getThrower`, as not all `EntityLivingBase` subclasses are `ProjectileSource`s.

**Pull request closed without merging**